Assistant Messages and Layout

Hamzali opened the pull request:
I have added `react-native-linear-gradient` for gradient background for message bubble and option buttons. If there is a better way to achieve gradient background please share.

I couldn't decide about the color scheme yet therefor, I left colors in place will update with your directions.

Used React Native's StyleSheet for native style reuse on components instead of plain objects therefore I needed to change functional styles with memoize.

Commit:
add message bubble, assistant layout

gwn commented on ios/Podfile.lock:
```diff
@@ -310,15 +312,18 @@ PODS:
   - RNKeychain (6.2.0):
     - React
-  - RNSVG (12.1.0):
+  - RNSVG (9.13.6):
```
Don't really care, but just curious. Any ideas why this change happened? Looks like the new `react-native-linear-gradient` dependency triggered a downgrade on this native module `RNSVG`, which is probably ok for now. Just wanted to hear from you if you have any deeper knowledge regarding this than me. But if not, no problem at all.

Hamzali replied:
reverted this change  by removing linear gradient

gwn commented on ios/ixoWallet.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>IDEDidComputeMac32BitWarning</key>
+	<true/>
+</dict>
+</plist>
```
Again, don't really care but just curious. Any ideas about what this file is?

Hamzali replied:
After some small research, it is a new config file to track env requirement for Xcode IDE.
![image](https://user-images.githubusercontent.com/11903062/108904183-40f0de80-762f-11eb-8cc4-85f59f740283.png)
https://developer.apple.com/library/archive/releasenotes/DeveloperTools/RN-Xcode/Chapters/Introduction.html#//apple_ref/doc/uid/TP40001051-CH1-DontLinkElementID_24

gwn replied:
wonderful thanks for the info

gwn commented on src/assistant/Assistant.js:
```diff
@@ -1,11 +1,23 @@
+// const {default: AssistantLogo} = require('../assets/top-assistant-icon.svg')
+// const {default: CloseIcon} = require('../assets/close-icon.svg')
+// const {default: ScanIcon} = require('../assets/qr-scan.svg')
+
```
Deleting is fine, no need to comment out. I guess you either forgot or left them because this patch is still a work in progress. Both are ok btw, no problem, just leaving a note.

gwn commented on src/assistant/Assistant.js:
```diff
@@ -1,11 +1,23 @@
+    CloseIcon = require('$/lib/icons/Close'),
+    AssistantIcon = require('$/lib/icons/Assistant'),
+    ArrowUpIcon = require('$/lib/icons/ArrowUp'),
+    {default: LinearGradient} = require('react-native-linear-gradient')
```
This having to import modules with the `default` keyword from time to time is really bothering me. This is because of my decision of using the CommonJS syntax over the ES Modules syntax, just because I prefer the client codebases to look like Node codebases as much as possible. But I can't deny this problem, although I can live with it. What about you? Any thoughts?

gwn replied:
We agreed to proceed with the import syntax in our phone conversation, so resolving.

gwn commented on src/lib/theme.js:
```diff
@@ -0,0 +1,28 @@
+const BASE_MEASURE = 8
+
+const spacing = (n) => n * BASE_MEASURE
+
+const shadow = () => ({
+    shadowColor: 'rgb(212, 221, 232)',
+    shadowOffset: {
+        width: 0,
+        height: 2,
+    },
+    shadowOpacity: 0.4,
+    shadowRadius: 15,
+    elevation: 1,
+})
+
+const colors = {
+    primary: '',
+    secondary: '',
+    background: '',
+    white: '',
+    black: '',
+}
+
+module.exports =  {
+    spacing,
+    shadow,
+    colors,
+}
```
Cool, exactly the way I imagined this.

gwn commented on src/lib/theme.js:
```diff
@@ -0,0 +1,28 @@
+
+const spacing = (n) => n * BASE_MEASURE
+
+const shadow = () => ({
```
I guess you're planning to parameterize this at some point, I assume you wouldn't implement this as a function otherwise. Am I right?

Hamzali replied:
Yes, it will have levels of shadow something like elevation level of the surface.

gwn replied:
ok that makes sense

gwn commented:
> I have added react-native-linear-gradient for gradient background for message bubble and option buttons. If there is a better way to achieve gradient background please share.

I don't know, but this looks fine. Thanks for asking.

> I couldn't decide about the color scheme yet therefor, I left colors in place will update with your directions.

All right, I don't know how to proceed there too for now, we can make a decision together in the following days.

> Used React Native's StyleSheet for native style reuse on components instead of plain objects therefore I needed to change functional styles with memoize.

I didn't know about "native style reuse". I wasn't seeing any benefit in using that `Stylesheet` thing so that's why I proceeded with plain objects in the first place. Interesting, so `Stylesheet` now does sound useful now, heh :) Anyway I'm ok with this.

gwn commented:
![ss](https://user-images.githubusercontent.com/1702469/108886969-adadae00-761a-11eb-92c2-88ed5923431c.png)

This is the way my Android build looks. The borders / outlines seem to be off. We agreed that Android is not our priority so an action is not really required at this point (if the iOS build is fine I mean). Just letting you know.

gwn commented:
I think this is a very good looking PR in general. Thank you.

gwn reviewed (changes requested):
I'm okay with everything except the button styles. I can see how you based the button designs on the 3rd screen of the design (which I'm attaching to this comment for reference), but that's not ok. There's an ambiguity in the design that some buttons are bordered & inline and others are not. I guess that the styling you went with were designed for wallet addresses as a special case which I think we should ignore for now.

My request is the restyling of the buttons to make them look like the bordered & inline buttons that can be seen in the first, 4th and 5th screens of the design.

![assistant](https://user-images.githubusercontent.com/1702469/109125209-d50c9400-775c-11eb-94dc-a31f414fe9df.jpg)

Hamzali commented:
updated the button styles;
![Screen Shot 2021-02-26 at 7 09 13 PM](https://user-images.githubusercontent.com/11903062/109324790-3704f080-7866-11eb-9a64-cf4c72de5d9b.png)

Hamzali commented:
option buttons also wrapped in a view, so it shows in one line if there is enough space for the buttons' content.

gwn committed:
Introduce a multiple selection widget

gwn committed:
Introduce date & date range picker widgets

I'm not very happy with the current library. It seems a bit slow
and also depends on Moment which is a big dependency. But since
it's simple and does the job properly, I'm ok with it for now.
Let's keep this in mind and replace it with a better performing
library (if exists) in the future.

gwn committed:
Introduce a generic entity filter widget

gwn committed:
Implement project filtering
